lua/null-ls/builtins/diagnostics.lua

```diff
@@ -5,6 +5,120 @@ local DIAGNOSTICS = methods.internal.DIAGNOSTICS
 
 local M = {}
 
+local default_severities = {
+    ["error"] = 1,
+    ["warning"] = 2,
+    ["information"] = 3,
+    ["hint"] = 4,
+}
+
+local default_json_attributes = {
+    row = "line",
+    col = "column",
+    end_row = "endLine",
+    end_col = "endColumn",
+    code = "ruleId",
+    severity = "level",
+    message = "message",
+}
+
+local make_attr_adapters = function(severities)
+    return {
+        severity = function(entries, _)
+            return severities[entries["severity"]]
+        end,
+        end_col = function(entries, line)
+            local end_col = entries["end_col"]
+            local quote = entries["quote"]
+            if end_col or not quote or not line then
+                return end_col
+            end
+
+            _, end_col = line:find(quote, 1, true)
+            return end_col
+        end,
+    }
+end
+
+local make_diagnostic = function(entries, defaults, attr_adapters, params)
+    if not (entries["row"] and entries["message"]) then
+        return nil
+    end
+
+    local content_line = params.content and params.content[tonumber(entries["row"])] or nil
+    for attr, adapter in pairs(attr_adapters) do
+        entries[attr] = adapter(entries, content_line) or entries[attr]
+    end
+
+    return vim.tbl_extend("keep", defaults, entries)
+end
+
+--- Parse a linter's output using a regex pattern
+-- @param pattern The regex pattern
+-- @param groups The groups defined by the pattern: {"line", "message", "col", ["end_col"], ["code"], ["severity"], ["quote"]}
+-- @param severities An optional table of severity overrides (see default_severities)
+-- @param defaults An optional table of diagnostic default values
+local from_pattern = function(pattern, groups, severities, defaults)
+    severities = vim.tbl_extend("force", default_severities, severities or {})
+    defaults = defaults or {}
+
+    local attr_adapters = make_attr_adapters(severities)
+    return function(line, params)
+        local results = { line:match(pattern) }
+        local entries = {}
+        for i, match in ipairs(results) do
+            entries[groups[i]] = match
+        end
+
+        return make_diagnostic(entries, defaults, attr_adapters, params)
+    end
+end
+
+--- Parse a linter's output using multiple regex patterns until one matches
+-- @param patterns The regex pattern list
+-- @param groups The groups list defined by the patterns
+-- @param severities An optional table of severity overrides (see default_severities)
+-- @param defaults An optional table of diagnostic default values
+local from_patterns = function(patterns, groups, severities, defaults)
+    return function()
+        for i, pattern in ipairs(patterns) do
+            local diagnostic = from_pattern(pattern, groups[i], severities, defaults)
+            if diagnostic then
+                return diagnostic
+            end
+        end
+        return nil
+    end
```

Comment:
I think this callback is missing the arguments it receives from `on_output`, right? I had to change it a bit to get it to work on my end:

```suggestion
    return function(line, params)
        for i, pattern in ipairs(patterns) do
            local diagnostic = from_pattern(pattern, groups[i], severities, defaults)(line, params)
            if diagnostic then
                return diagnostic
            end
        end
        return nil
    end
```